dcs/coalition.py

```diff
@@ -89,172 +89,188 @@ def get_name(mission: "Mission", name: str) -> str:
-
-                    Coalition._import_moving_point(mission, helicopter_group, pgroup)
-
-                    # units
-                    for imp_unit_idx in pgroup["units"]:
-                        imp_unit = pgroup["units"][imp_unit_idx]
-                        heli = Helicopter(
-                            mission.terrain,
-                            _id=imp_unit["unitId"],
-                            name=self.get_name(mission, imp_unit["name"]),
-                            _type=helicopters.helicopter_map[imp_unit["type"]],
-                            _country=_country)
-                        heli.load_from_dict(imp_unit)
-
-                        if _country.reserve_onboard_num(heli.onboard_num):
-                            msg = "{c} Helicopter '{h}' already using tail number: {t}".format(
-                                c=self.name.upper(), h=heli.name, t=heli.onboard_num)
-                            status.append(StatusMessage(msg, MessageType.ONBOARD_NUM_DUPLICATE, MessageSeverity.WARN))
-                            print("WARN:", msg, file=sys.stderr)
-                        status += self._park_unit_on_airport(mission, helicopter_group, heli)
-
-                        mission.current_unit_id = max(mission.current_unit_id, heli.id)
-                        helicopter_group.add_unit(heli)
-
-                    # check runway start
-                    # if helicopter_group.points[0].airdrome_id is not None and helicopter_group.units[0].parking is None:
-                    #     airport = mission.terrain.airport_by_id(helicopter_group.points[0].airdrome_id)
-                    #     airport.occupy_runway(helicopter_group)
-                    _country.add_helicopter_group(helicopter_group)
-
-            if "static" in imp_country:
-                for sgroup_idx in imp_country["static"]["group"]:
-                    sgroup = imp_country["static"]["group"][sgroup_idx]
-                    static_group = unitgroup.StaticGroup(sgroup["groupId"],
-                                                         self.get_name(mission, sgroup["name"]))
-                    static_group.load_from_dict(sgroup, mission.terrain)
-                    mission.current_group_id = max(mission.current_group_id, static_group.id)
-
-                    Coalition._import_static_point(mission, static_group, sgroup)
-
-                    # units
-                    for imp_unit_idx in sgroup["units"]:
-                        imp_unit = sgroup["units"][imp_unit_idx]
-                        static: Static
-                        if imp_unit["type"] == "FARP":
-                            static = FARP(
+
+        # add countries to coalition even if they have no units
+        for idx, country_id in countries_in_coalition.items():
+            _country = countries.get_by_id(country_id)
+            if country_has_units(country_id, d["country"]):
+                country_idx = get_country_idx(country_id, d["country"])
```

Comment:
You've got a `has` and `get` that perform identical operations with the only difference being the returned value. You can rewrite by replacing `country_has_units` and `get_country_idx`

```python
def country_dict_from_id(country_id: int, d: Dict[int, Any]) -> Optional[dict[str, Any]]:
    # Same implementation as get_country_idx, more or less
```

(no idea if that name is correct or not, pick whatever makes sense, that's just an example)

And then rewrite the loop as:

```python
for country_id in countries_in_coalition.values():
    self.add_country(self.countries.get_by_id(country_id)
    imp_country = country_dict_from_id(country_id, d["country"])
    if imp_country is None:
        continue

    # rest of the function as before, though the last line moved up so it wouldn't be skipped
```

Less code and runs faster (it doesn't search for the same value 3 times). (It'd be even easier if the miz table was keyed by country ID rather than by index, but whatever, can't fix that)

Alternatively (and probably better?), put this `for` back the way it was, but move the `self.add_country()` call out into a separate

```python
for country_id in countries_in_coalition.values():
    self.add_country(country_id)
```

outside this loop and then you can remove the lookup functions entirely.

Reply:
@DanAlbert - I like the idea of keeping the old code for what it did, then adding at the end of the function a simple loop loop that checks if a country is in the coalitions, skip it (since the orig code processed it), otherwise add it.